libbpf-tools/execsnoop.c

```diff
@@ -0,0 +1,338 @@
+    { "line", 'l', "LINE", 0, "only print commands where arg contains this line"},
+    { "print-uid", 'U', NULL, 0, "print UID column"},
+    { "max-args", MAX_ARGS_KEY, "MAX_ARGS", 0,
+		"maximum number of arguments parsed and displayed, defaults to 20"},
+    { "verbose", 'v', NULL, 0, "Verbose debug output" },
+    {},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+    long int uid, max_args;
+
+    switch (key) {
+    case 'h':
+        argp_usage(state);
+        break;
+    case 'T':
+        env.time = true;
+        break;
+    case 't':
+        env.timestamp = true;
+        break;
+    case 'x':
+        env.fails = true;
+        break;
+    case 'u':
+        errno = 0;
+        uid = strtol(arg, NULL, 10);
+        if (errno || uid < 0 || uid >= INVALID_UID) {
+            fprintf(stderr, "Invalid UID %s\n", arg);
+            argp_usage(state);
+        }
+        env.uid = uid;
+        break;
+    case 'q':
+        env.quote = true;
+        break;
+    case 'n':
+		env.name = arg;
+        break;
+    case 'l':
+		env.line = arg;
+        break;
+    case 'U':
+        env.print_uid = true;
+        break;
+    case 'v':
+        env.verbose = true;
+        break;
+	case MAX_ARGS_KEY:
+        errno = 0;
+        max_args = strtol(arg, NULL, 10);
+        if (errno || max_args < 1 || max_args > TOTAL_MAX_ARGS) {
+            fprintf(stderr, "Invalid MAX_ARGS %s, should be in [1, %d] range\n",
+					arg, TOTAL_MAX_ARGS);
+
+            argp_usage(state);
+        }
+		env.max_args = max_args;
+		break;
+    default:
+        return ARGP_ERR_UNKNOWN;
+    }
+    return 0;
+}
+
+static int bump_memlock_rlimit(void)
+{
+    struct rlimit rlim_new = {
+        .rlim_cur   = RLIM_INFINITY,
+        .rlim_max   = RLIM_INFINITY,
+    };
+
+    return setrlimit(RLIMIT_MEMLOCK, &rlim_new);
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+            const char *format, va_list args)
+{
+    if (level == LIBBPF_DEBUG && !env.verbose)
+        return 0;
+    return vfprintf(stderr, format, args);
+}
+
+static void time_since_start(char *str, size_t size)
+{
+	int ret;
+	long nsec, sec;
+	static struct timespec cur_time;
+
+    clock_gettime(CLOCK_MONOTONIC, &cur_time);
+	nsec = cur_time.tv_nsec - start_time.tv_nsec;
+	sec = cur_time.tv_sec - start_time.tv_sec;
+	if (nsec < 0) {
+		nsec += NSEC_PER_SEC;
+		sec--;
+	}
+	/* rounded division */
+	nsec = (nsec + (NSEC_PRECISION / 2)) / NSEC_PRECISION;
+	ret = snprintf(str, size, "%ld.%ld", sec, nsec);
+	if (ret < 0) {
+		str[size - 1] = '\0';
+	}
+}
+
+static void print_args(const struct event *e, bool quote)
+{
+	for (int i = 0; i < e->args_count && i < TOTAL_MAX_ARGS; i++) {
+        const char *c = &(e->args[i * ARGSIZE]);
+        int j = 0;
+		if (i > 0) {
+			putchar(' ');
+		}
+		/* arguments should be quoted */
+		if (env.quote) {
+			putchar('"');
+			while(*c && j < ARGSIZE) {
+				switch(*c) {
+					case '"':
```

Comment:
Python version [handle only](https://github.com/iovisor/bcc/blob/master/tools/execsnoop.py#L288) `"`, but I think it's better if quote behavior will be more like python `repr()`.

Reply:
extract this into function, please